### ISSUE TITLE
Support for configurable Keep-Alive for both frontend and backend 

### ISSUE THREAD

**Issue by bgv** [containous/traefik#727]
Hi,
I want to propose adding new configuration options prompt by #718.

While certainly its nice to have Keep-Allive from end to end sometimes we don't have full control over the applications we deploy.

In some situations it's good to have more granular control over the connection, so for example if we use the application just to print few bits of information we don't need the connection staying open for another 30 seconds.

For example when creating new service with docker swarm we can just add new label for this particular frontend: `--label traefik.frontend.rule=Client-KeepAlive:5,Backend-KeepAlive:0` which will tell traefik to keep the connection to client  browser open for 5 sec but close connection to backend after the response is finished.

So the full service command becomes:

``` bash
docker service create --name whoami --replicas 3 --label traefik.port=80 --label traefik.frontend.rule=Client-KeepAlive:5,Backend-KeepAlive:0 --network frontend emilevauge/whoami
```

I don't think this is urgent feature but certainly good to have.
Cheers.

**Comment by jsravn** [comment]
I think this feature is pretty urgent. It means traefik will generate errors for non-idempotent requests - go http client only retries closed idle connections for idempotent requests. Only way to prevent this is to set the forwarder's transport idle timeout to less than our server timeouts. In golang 1.6 this isn't straightforward, while 1.7 has added the idle timeout that can be used for this. I'm not sure how easy it will be to do this, it probably requires a change to the oxy library that traefik uses.

**Comment by 27Bslash6** [comment]
Adding my reasoning for this request: optimal configuration for Akamai <> origin is keepalive of 301 seconds (https://community.akamai.com/customers/s/question/0D50f00005RtqgwCAB/default-pconn-timeout?language=en_US) but we don't want Traefik <> backend service connections persisting this long

**Comment by stealthybox** [comment]
+1
It may be difficult/impossible to disable keep-alive support on a backend server
(particularly if the servers are part of a third party application).

#### Common use-case:
It's difficult to do zero-downtime rolling deploys of NGINX pods if keep-alive is enabled.
NGINX can't drain keep-alive connections during a graceful shutdown so it drops in-flight requests:
- https://github.com/kubernetes/contrib/issues/1123
- https://trac.nginx.org/nginx/ticket/1022

Say we're using Traefik to load-balance some NGINX servers;
It would be convenient to work around this issue by configuring the Traefik to not keep-alive those backend connections.

**Comment by pmontepagano** [comment]
So... currently there is no way to completely disable keepalives between Traefik and a backend?

**Comment by glebsam** [comment]
> So... currently there is no way to completely disable keepalives between Traefik and a backend?

There is: https://github.com/containous/traefik/issues/3237#issuecomment-514178590

**Comment by bgv** [comment]
#4983 is merged, and is in v2.0.0 (still beta):

- https://github.com/Wikia/traefik/blob/a79b942edce4fc08ef744e291bb117284147d33a/docs/content/reference/static-configuration/file.toml#L12
- https://github.com/Wikia/traefik/blob/a79b942edce4fc08ef744e291bb117284147d33a/docs/content/reference/static-configuration/cli.txt#L480

Closing this issue now, many thanks to everyone involved.
